### PR TITLE
WA-403: Implement new transaction flow UX/UI

### PR DESCRIPTION
**Pull request by rmeissner**
Closes WA-403.

Changes proposed in this pull request:
- Apply new design for transaction flow

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/create/CreateAssetTransferActivity.kt [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+            is CreateAssetTransferContract.ViewUpdate.TokenInfo -> {
+                update.value.token.name?.let{
+                    layout_create_asset_transfer_title.text = getString(R.string.transfer_x, it)
+                }
+                layout_create_asset_transfer_safe_balance.text = update.value.displayString()
+                layout_create_asset_transfer_input_label.text = update.value.token.symbol ?: "???"
+            }
+            is CreateAssetTransferContract.ViewUpdate.InvalidInput -> {
+                layout_create_asset_transfer_input_value.setTextColor(
+                    ContextCompat.getColor(
```
Comment (fmrsabino) [inline app/src/main/java/pm/gnosis/heimdall/ui/transactions/create/CreateAssetTransferActivity.kt]: `getColorCompat`

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/ReviewTransactionViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+class ReviewTransactionViewModel @Inject constructor(
+    private val addressHelper: AddressHelper,
+    private val executionRepository: TransactionExecutionRepository,
+    private val signaturePushRepository: SignaturePushRepository,
+    private val signatureStore: SignatureStore,
+    private val tokenRepository: TokenRepository
+) : ReviewTransactionContract() {
+
+    private var safe: Solidity.Address? = null
```
Comment (fmrsabino) [inline app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/ReviewTransactionViewModel.kt]: I'd declare this as a `lateinit` var. The exception thrown by not initializing the field would be caught up in an `Observable` and would remove a lot of `!!`

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import pm.gnosis.heimdall.data.repositories.TransactionData
+import pm.gnosis.heimdall.data.repositories.TransactionExecutionRepository
+import pm.gnosis.heimdall.data.repositories.models.ERC20Token
+import pm.gnosis.heimdall.data.repositories.models.SafeTransaction
+import pm.gnosis.heimdall.helpers.AddressHelper
+import pm.gnosis.heimdall.ui.transactions.builder.AssetTransferTransactionBuilder
+import pm.gnosis.heimdall.ui.transactions.review.TransactionInfoViewHolder
+import pm.gnosis.model.Solidity
+import pm.gnosis.models.Transaction
+import pm.gnosis.models.Wei
```
Comment (fmrsabino) [inline app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt]: Unused imports

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        if (data.token == ERC20Token.ETHER_TOKEN.address)
+            Single.just(ERC20Token.ETHER_TOKEN)
+        else
+            tokenRepository.loadToken(data.token)
+                .onErrorResumeNext {
+                    tokenRepository.loadTokenInfo(data.token).firstOrError()
+                }
+
+    private fun setupSafeInfo() {
+        val view = view ?: return
```
Comment (fmrsabino) [inline app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt]: These checks on every method are not the best 🤔

File: app/src/main/java/pm/gnosis/heimdall/ui/security/unlock/UnlockDialog.kt [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        layout_unlock_dialog_password_input.disableAccessibility()
+
+        disposables += layout_unlock_dialog_password_input.editorActions()
+            .filter { it == EditorInfo.IME_ACTION_DONE || it == EditorInfo.IME_NULL }
+            .flatMap { viewModel.unlock(layout_unlock_dialog_password_input.text.toString()) }
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeForResult(onNext = ::onState, onError = ::onStateCheckError)
+
+        layout_unlock_dialog_password_input.keyPreImeListener = { keyCode, event ->
+            if (keyCode == KeyEvent.KEYCODE_BACK) {
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/ui/security/unlock/UnlockDialog.kt]: no that is only for the dialog, but that is the flow specified in the designs

File: app/src/main/java/pm/gnosis/heimdall/ui/security/unlock/UnlockDialog.kt [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeForResult(onNext = ::onFingerprintResult, onError = ::onFingerprintError)
+    }
+
+    private fun togglePasswordInput(visible: Boolean) {
+        layout_unlock_dialog_fingerprint.visible(!visible)
+        layout_unlock_dialog_password_input.visible(visible)
+        layout_unlock_dialog_progress.visible(false)
+        if (visible) layout_unlock_dialog_password_input.showKeyboardForView()
+        else (context?.getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager)?.hideSoftInputFromWindow(
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/ui/security/unlock/UnlockDialog.kt]: we have something like that in svalinn but it requires the Activity, we can add that there later

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/create/CreateAssetTransferActivity.kt (hunk 1 above)
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/ui/transactions/create/CreateAssetTransferActivity.kt]: will change with the next PR everywhere

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/ReviewTransactionViewModel.kt (hunk 2 above)
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/ReviewTransactionViewModel.kt]: will change with the next PR

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt (hunk 3 above)
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt]: will change with the next PR

File: app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt (hunk 4 above)
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/ui/transactions/review/viewholders/AssetTransferViewHolder.kt]: why "not the best" ... they don't look nice I agree. I could also use lateinit, but that would potentially crash the app. I will change some of them to use `apply {}` instead ... looks nicer

File: app/src/main/java/pm/gnosis/heimdall/services/HeimdallFirebaseService.kt [hunk 7]
```diff
@@ -70,8 +70,8 @@ class HeimdallFirebaseService : FirebaseMessagingService() {
     }
 
     private fun showNotification(signRequest: GnoSafeUrlParser.Parsed.SignRequest) {
-        val safeTransaction = SafeTransaction(signRequest.transaction, TransactionRepository.Operation.values()[signRequest.operation])
-        val intent = SignTransactionActivity.createIntent(this, signRequest.safe, safeTransaction, true)
+        val safeTransaction = SafeTransaction(signRequest.transaction, TransactionExecutionRepository.Operation.values()[signRequest.operation])
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/services/HeimdallFirebaseService.kt]: Fixed in a future PR

**Pull request merged**